zanata-war/src/main/java/org/zanata/action/ZanataRoleAction.java

```diff
@@ -0,0 +1,131 @@
+package org.zanata.action;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.jboss.seam.annotations.Begin;
+import org.jboss.seam.annotations.Create;
+import org.jboss.seam.annotations.In;
+import org.jboss.seam.annotations.Install;
+import org.jboss.seam.annotations.Name;
+import org.jboss.seam.annotations.Scope;
+import org.jboss.seam.core.Conversation;
+import org.zanata.security.ZanataIdentity;
+import org.zanata.seam.security.ZanataJpaIdentityStore;
+import org.zanata.security.annotations.CheckLoggedIn;
+import org.zanata.security.annotations.CheckRole;
+import org.zanata.security.annotations.ZanataSecured;
+
+import static org.jboss.seam.ScopeType.CONVERSATION;
+import static org.jboss.seam.annotations.Install.APPLICATION;
+
+/**
+ * @author Patrick Huang
+ *         <a href="mailto:pahuang@redhat.com">pahuang@redhat.com</a>
+ */
+@Name("zanataRoleAction")
+@Scope(CONVERSATION)
+@Install(precedence = APPLICATION)
+@ZanataSecured
+@CheckLoggedIn
+@CheckRole("admin")
+public class ZanataRoleAction implements Serializable {
```

Comment:
On 2015-07-31 15:54, Patrick Huang wrote:
> In zanata-war/src/main/java/org/zanata/action/ZanataRoleAction.java
> <https://github.com/zanata/zanata-server/pull/908#discussion_r35948181>:
> 
>> +import org.zanata.security.annotations.ZanataSecured;
>> +
>> +import static org.jboss.seam.ScopeType.CONVERSATION;
>> +import static org.jboss.seam.annotations.Install.APPLICATION;
>> +
>> +/**
>> + * @author Patrick Huang
>> + *         <a href="mailto:pahuang@redhat.com">pahuang@redhat.com</a>
>> + */
>> +@Name("zanataRoleAction")
>> +@Scope(CONVERSATION)
>> +@Install(precedence = APPLICATION)
>> +@ZanataSecured
>> +@CheckLoggedIn
>> +@CheckRole("admin")
>> +public class ZanataRoleAction implements Serializable {
> 
> So what's your suggestion?

I suggest we keep it as close to the Seam status quo as possible, for now.

> Should I remove that check from
> ZanataRoleAction, ZanataRoleSearch and ZanataUserAction,
> ZanataUserSearch?

Unless they or their Seam relatives had admin checks before, yes.

> They are secured by /admin/* in pages.xml from being
> accessed.

And that's still fine for the moment.  It's just an extra layer of
defence for what we currently consider "admin" pages.

As your TODO in pages.xml says, we will need a CDI option for this.
That might turn out to be putting @CheckRole("admin") on a lot of beans
like ZanataRoleAction, but I hope it doesn't come to that.  Not yet, anyway.


> They are also secured by identity manager. and of course the
> SmartEntityListener at the lowest level.

I think SmartEntitySecurityListener and ZanataIdentity should perhaps
check the same permissions[*] (based on EntityAction, not strings like
"create" and "update"), but I don't want to take on refactoring Seam
just yet, so let's just keep it the same as before.

[*] Then again, checking the same permissions over and over again is a
bit pointless and high maintenance.

I suppose as a general rule, we really want to check permissions at the
lowest practical abstraction level, to avoid the danger of extra code
paths which fail to check permissions.  For basic CRUD stuff, that
probably means checking at the entity level.  However, for the sake of
being user friendly, we need to perform early checks of the same
permissions at the front end (UI, perhaps REST too), so that we can turn
off (grey out or hide) disallowed features in the UI.

Carlos, any advice on this?


> If we want to change the
> permission rule, all these layers needs to change.

Only the layers which check explicitly for the admin role.  Otherwise,
we just have to make sure the relevant permissions are granted to the
right users.


-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat